### DISCUSSION TITLE
Does wfb-ng & wfb-ng-osd support iNav/betaFlight？

### DISCUSSION THREAD

**Discussion by lida2003** [svpcom/wfb-ng#290] [Questions] [answered]
From readme， we know that it should support PX4/Ardupilot.
does wfb-ng & wfb-ng-osd support iNav/betaFlight？ If it supports, how to configure?

**Comment by svpcom** [comment, marked as answer]
Out of box it supports any autopilot with mavlink. If inav works with mavlink telemetry it will work. If it use different telemetry protocol then you need to write it parser (which will read bytes from uart and parse them to packets) and add it to wfb-ng

**Reply by lida2003** [reply]
inav indeed support some mavlink protocols. I didn't take a close look at this. Did your try osd on those firmware(inav/betaflight)?

**Reply by svpcom** [reply]
No. I've only test PX4 and ardupilot

**Comment by whoim2** [comment]
This is utility for mapping wfb processes to serial, this version "transparent" for msp one side (only air to ground), try compile it for you ground

https://github.com/OpenIPC/sandbox-fpv/tree/master/user_TipoMan

**Reply by lida2003** [reply]
it seems broken.  @whoim2 
```
pi@raspberrypi3b:~ $ tar -xf mavfwd_mavlink2.tar
tar: This does not look like a tar archive
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
pi@raspberrypi3b:~ $ ls -lh mavfwd_mavlink2.tar
-rwx------ 1 pi pi 204K Mar 21 09:12 mavfwd_mavlink2.tar
```

**Reply by whoim2** [reply]
> it seems broken. @whoim2
> 
> ```
> pi@raspberrypi3b:~ $ tar -xf mavfwd_mavlink2.tar
> tar: This does not look like a tar archive
> tar: Skipping to next header
> tar: Exiting with failure status due to previous errors
> pi@raspberrypi3b:~ $ ls -lh mavfwd_mavlink2.tar
> -rwx------ 1 pi pi 204K Mar 21 09:12 mavfwd_mavlink2.tar
> ```

Open this 7zip on win

**Reply by lida2003** [reply]
I can't open this either

![图片](https://user-images.githubusercontent.com/1296171/226557063-4d680c58-2b75-4a9a-be1c-faed9b788f72.png)

**Reply by svpcom** [reply]
User `tar xvf mavfwd_mavlink2.tar`. Ensure that you download attachment binary, not github html page

**Reply by lida2003** [reply]
hummm, it might be html issue. I have got the code.

![图片](https://user-images.githubusercontent.com/1296171/226564856-f1ec68ae-d531-40e0-9baa-9d5fe6a172e7.png)

**Comment by tipoman9** [comment]
Hi all,
The original version of mavfwd tries to parse the mavlink packets, which I do not think is needed for a simple forwarding.
But it splits them only by 0xFE.

> The major version can be determined from the packet start marker byte:
> MAVLink 1: 0xFE
> MAVLink 2: 0xFD

[https://mavlink.io/en/guide/mavlink_version.html](url)

So I simply commented this parsing and it worked with INAV and wfb_ng.
But I think the code as it is now is not reliable enough regardless of the separator and prone to miss packets.
It assumes that data from serial port will arrive in frames, which is not guaranteed.
The read from the serial port stream does not guarantee that you well get a complete packet.
So it has to be improved a little to parse packets. When this is done, it would be very easy to add packet aggregation, based on minimum data contents size and max latency for example.
